File: auth-service/src/main/java/com/user/login/service/ProductService.java
package com.user.login.service;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

import com.user.login.client.InventoryServiceClient;
import com.user.login.dto.InventoryResponse;
import com.user.login.dto.ProductRequest;
import com.user.login.dto.ProductResponse;
import com.user.login.entity.Product;
import com.user.login.repository.ProductRepository;

@Service
public class ProductService {

    private final ProductRepository repository;
    
    private final InventoryServiceClient inventoryServiceClient;

    public ProductService(ProductRepository repository,InventoryServiceClient inventoryServiceClient) {
        this.repository = repository;
        this.inventoryServiceClient = inventoryServiceClient;
    }

    /*
     * Add Product
     */
    public ProductResponse saveProduct(ProductRequest request) {

        Product product = new Product();

        product.setProductName(request.getProductName());
        product.setCategory(request.getCategory());
        product.setPrice(request.getPrice());
        product.setDescription(request.getDescription());

        Product savedProduct = repository.save(product);

        return mapToResponse(savedProduct);
    }

    /*
     * Get All Products
     */
    public List<ProductResponse> getAllProducts() {

        // 1. Get all products from Product table
        List<Product> products = repository.findAll();

        // 2. Get all inventory in ONE call
        List<InventoryResponse> inventories =
                inventoryServiceClient.getAllInventory();

        // 3. Convert inventory list to Map
        java.util.Map<Long, Integer> quantityMap =
                inventories.stream()
                        .collect(Collectors.toMap(
                                InventoryResponse::getProductId,
                                InventoryResponse::getAvailableQuantity
                        ));

        // 4. Combine Product + Inventory
        return products.stream()
                .map(product -> {

                    Integer quantity =
                            quantityMap.getOrDefault(
                                    product.getId(),
                                    0
                            );
                    String productImage =
                            "/products/" + product.getId() + ".jpg";

                    return new ProductResponse(

                            product.getId(),

                            product.getProductName(),

                            product.getCategory(),

                            product.getPrice(),

                            quantity,

                            product.getDescription(),
                            productImage
                    );

                })
                .collect(Collectors.toList());
    }

    /*
     * Get Product By Id
     */
    public ProductResponse getProductById(Long id) {

        Product product = repository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Product Not Found"));

        return mapToResponse(product);

    }

    /*
     * Update Product
     */
    public ProductResponse updateProduct(Long id,
                                         ProductRequest request) {

        Product product = repository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Product Not Found"));

        product.setProductName(request.getProductName());
        product.setCategory(request.getCategory());
        product.setPrice(request.getPrice());
        product.setDescription(request.getDescription());

        Product updated = repository.save(product);

        return mapToResponse(updated);

    }

    /*
     * Delete Product
     */
    public void deleteProduct(Long id) {

        repository.deleteById(id);

    }

    /*
     * Common Mapping Method
     */
    /*
     * Common Mapping Method
     * Product + Inventory
     */
    private ProductResponse mapToResponse(Product product) {
    	
    	 String productImage =
    	            "/products/" + product.getId() + ".jpg";

        InventoryResponse inventory =
                inventoryServiceClient.getInventoryByProductId(
                        product.getId()
                );

        Integer quantity = 0;

        if (inventory != null) {
            quantity = inventory.getAvailableQuantity();
        }

        return new ProductResponse(

                product.getId(),

                product.getProductName(),

                product.getCategory(),

                product.getPrice(),

                quantity,

                product.getDescription(),
                productImage
        );
    }
    
    /*
     * Search Products
     */
    public List<ProductResponse> searchProducts(String keyword) {

        return repository
                .findByProductNameContainingIgnoreCaseOrCategoryContainingIgnoreCaseOrDescriptionContainingIgnoreCase(
                        keyword,
                        keyword,
                        keyword)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());

    }

    /*
     * Get Products By Category
     */
    public List<ProductResponse> getProductsByCategory(String category) {

        return repository
                .findByCategoryIgnoreCase(category)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());

    }
    
    
    }

File: jwt-auth-ui/src/services/productService.js
import axios from "axios";

import { getToken } from "../utils/tokenStorage";

const API_BASE_URL = "http://localhost:8080/api/products";

/*
 * Common Request Configuration
 */
const getConfig = () => {

    return {

        headers: {

            Authorization: `Bearer ${getToken()}`,

            "Content-Type": "application/json"

        }

    };

};

/*
 * Get All Products
 */
export const getProducts = async () => {

    const response = await axios.get(

        API_BASE_URL,

        getConfig()

    );

    return response.data;

};

/*
 * Get Product By Id
 */
export const getProductById = async (id) => {

    const response = await axios.get(

        `${API_BASE_URL}/${id}`,

        getConfig()

    );

    return response.data;

};

/*
 * Save Product
 */
export const saveProduct = async (product) => {

    const response = await axios.post(

        API_BASE_URL,

        product,

        getConfig()

    );

    return response.data;

};

/*
 * Update Product
 */
export const updateProduct = async (id, product) => {

    const response = await axios.put(

        `${API_BASE_URL}/${id}`,

        product,

        getConfig()

    );

    return response.data;

};

/*
 * Delete Product
 */
export const deleteProduct = async (id) => {

    const response = await axios.delete(

        `${API_BASE_URL}/${id}`,

        getConfig()

    );

    return response.data;

};

/*
 * Search Products
 */
export const searchProducts = async (keyword) => {

    const response = await axios.get(

        `${API_BASE_URL}/search?keyword=${encodeURIComponent(keyword)}`,

        getConfig()

    );

    return response.data;

};

/*
 * Get Products By Category
 */
export const getProductsByCategory = async (category) => {

    const response = await axios.get(

        `${API_BASE_URL}/category/${encodeURIComponent(category)}`,

        getConfig()

    );

    return response.data;

};
/*
 * Product Suggestions
 */
export const getSuggestions = async (keyword) => {

    const response = await axios.get(

        `${API_BASE_URL}/suggest?keyword=${encodeURIComponent(keyword)}`,

        getConfig()

    );

    return response.data;

};